phpBB/phpbb/ucp/controller/reset_password.php

```diff
@@ -0,0 +1,436 @@
+			 */
+			$vars = [
+				'email',
+				'username',
+				'sql_array',
+			];
+			extract($this->dispatcher->trigger_event('core.ucp_remind_modify_select_sql', compact($vars)));
+
+			$sql = $this->db->sql_build_query('SELECT', $sql_array);
+			$result = $this->db->sql_query_limit($sql, 2); // don't waste resources on more rows than we need
+			$rowset = $this->db->sql_fetchrowset($result);
+
+			if (count($rowset) > 1)
+			{
+				$this->db->sql_freeresult($result);
+
+				$this->template->assign_vars([
+					'USERNAME_REQUIRED'	=> true,
+					'EMAIL'				=> $email,
+				]);
+			}
+			else
+			{
+				$message = $this->language->lang('PASSWORD_RESET_LINK_SENT') . '<br /><br />' . $this->language->lang('RETURN_INDEX', '<a href="' . append_sid("{$this->root_path}index.{$this->php_ext}") . '">', '</a>');
+
+				$user_row = empty($rowset) ? [] : $rowset[0];
+				$this->db->sql_freeresult($result);
+
+				if (!$user_row)
+				{
+					return $this->helper->message($message);
+				}
+
+				if ($user_row['user_type'] == USER_IGNORE || $user_row['user_type'] == USER_INACTIVE)
+				{
+					return $this->helper->message($message);
+				}
+
+				// Do not create multiple valid reset tokens
+				if (!empty($user_row['reset_token']) && (int) $user_row['reset_token_expiration'] >= time())
+				{
+					return $this->helper->message($message);
+				}
+
+				// Check users permissions
+				$auth2 = new auth();
+				$auth2->acl($user_row);
+
+				if (!$auth2->acl_get('u_chgpasswd'))
+				{
+					return $this->helper->message($message);
+				}
+
+				// Generate reset token
+				$reset_token = strtolower(gen_rand_string(32));
+
+				$sql_ary = [
+					'reset_token'				=> $reset_token,
+					'reset_token_expiration'	=> strtotime('+1 day'),
+				];
+
+				$sql = 'UPDATE ' . $this->tables['users'] . '
+					SET ' . $this->db->sql_build_array('UPDATE', $sql_ary) . '
+					WHERE user_id = ' . $user_row['user_id'];
+				$this->db->sql_query($sql);
+
+				if (!class_exists('messenger'))
+				{
+					include($this->root_path . 'includes/functions_messenger.' . $this->php_ext);
+				}
+
+				/** @var \messenger $messenger */
+				$messenger = new \messenger(false);
+
+				$messenger->template('user_forgot_password', $user_row['user_lang']);
+
+				$messenger->set_addresses($user_row);
+
+				$messenger->anti_abuse_headers($this->config, $this->user);
+
+				$messenger->assign_vars([
+						'USERNAME'			=> htmlspecialchars_decode($user_row['username']),
+						'U_RESET_PASSWORD'	=> generate_board_url(true) . $this->helper->route('phpbb_ucp_reset_password_controller', [
+							'u'		=> $user_row['user_id'],
+							'token'	=> $reset_token,
+						], false)
+				]);
+
+				$messenger->send($user_row['user_notify_type']);
+
+				return $this->helper->message($message);
+			}
+		}
+
+		$this->template->assign_vars([
+			'USERNAME'			=> $username,
+			'EMAIL'				=> $email,
+			'S_PROFILE_ACTION'	=> $this->helper->route('phpbb_ucp_forgot_password_controller'),
```

Comment:
Shouldn't `<form>` action attributes not just be considered as url's and thus prefixed with `U_`?

Reply:
This is taken directly from the previous code but yeah, it's very inconsistent with the intended use of the S_ prefix.